### PR TITLE
Custom data

### PR DESCRIPTION
**Pull request by localvar**
refactor custom data, now Easegress & EaseMesh uses the same implementation

### CODE REVIEW

File: doc/reference/customdata.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+The YAML example below defines a CustomDataKind:
+
+```yaml
+name: kind1
+idField: name
+jsonSchema:
+  type: object
+  properties:
+    name:
+	  type: string
```
Comment (samutamm) [inline doc/reference/customdata.md:18]:
```suggestion
      type: string
```

File: doc/reference/customdata.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	  required: true
```
Comment (samutamm) [inline doc/reference/customdata.md:19]:
```suggestion
      required: true
```

File: doc/reference/customdata.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+The `name` field is required, it is the kind name of custom data items of this kind.
+The `idField` is optional, and its default value is `name`, the value of this field of a data item is used as its identifier.
+The `jsonSchema` is optional, if provided, data items of this kind will be validated against this [JSON Schema](http://json-schema.org/).
+
+## CustomData
+
+CustomData is a map, the keys of this map must be strings while the values can be any valid JSON values, but when value is also a map, its keys must be strings too.
+
```
Comment (samutamm) [inline doc/reference/customdata.md:29]: Probably very rare edge case but what about map containing slice/array of maps? In
`map[string][[]map[string]sometype]` I guess `sometype` can also only be `string` but not `slice` or `map`?

Is it correct to say:
```suggestion
In other words, nested CustomData objects can have at maximum the depth of two levels.
```

File: doc/reference/customdata.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+* **Delete all CustomData of a kind**
+	* **URL**: http://{ip}:{port}/apis/v1/customdata/{kind name}
+	* **Method**: DELETE
+
+* **Bulk update**
+	* **URL**: http://{ip}:{port}/apis/v1/customdata/{kind name}/items
+	* **Method**: POST
+	* **Body**: A change request in YAML, as defined below.
+
+		```yaml
```
Comment (samutamm) [inline doc/reference/customdata.md:96]:
```suggestion
```yaml
```

File: doc/reference/customdata.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		rebuild: false
```
Comment (samutamm) [inline doc/reference/customdata.md:97]:
```suggestion
rebuild: false
```

File: doc/reference/customdata.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		delete: [data1, data2]
```
Comment (samutamm) [inline doc/reference/customdata.md:98]:
```suggestion
delete: [data1, data2]
```

File: doc/reference/customdata.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+		list:
```
Comment (samutamm) [inline doc/reference/customdata.md:99]:
```suggestion
list:
```

File: doc/reference/customdata.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		- name: data3
```
Comment (samutamm) [inline doc/reference/customdata.md:100]:
```suggestion
- name: data3
```

File: doc/reference/customdata.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+			field1: 12
```
Comment (samutamm) [inline doc/reference/customdata.md:101]:
```suggestion
  field1: 12
```

File: doc/reference/customdata.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		- name: data4
```
Comment (samutamm) [inline doc/reference/customdata.md:102]:
```suggestion
- name: data4
```

File: doc/reference/customdata.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+			field1: foo
```
Comment (samutamm) [inline doc/reference/customdata.md:103]:
```suggestion
  field1: foo
```

File: pkg/cluster/clustertest/cluster.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	return 0
+}
+
+// Del implements STM.Del
+func (stm *MockedSTM) Del(key string) {
+	if stm.MockedRev != nil {
+		stm.MockedDel(key)
+	}
+}
```
Comment (samutamm) [inline pkg/cluster/clustertest/cluster.go:264]: Just noting here that after this PR this `MockedCluster` could be used in `pkg/filter/headerlookup/headerlookup_test.go` and some other Filter tests too to make the tests run faster.

File: doc/reference/customdata.md (hunk 3 above)
Comment (localvar) [reply doc/reference/customdata.md:29]: No, there's no limitation for the depth.
I mean if there are nested maps (either as a value or as an array item), the keys of the map must be strings.

Comment (samutamm) [reply doc/reference/customdata.md:29]: Ah Ok I get it now.

**Review by samutamm — APPROVED** [review]

**Review by suchen-sci — APPROVED** [review]

File: pkg/cluster/clustertest/cluster.go (hunk 12 above)
Comment (samutamm) [reply pkg/cluster/clustertest/cluster.go:264]: https://github.com/megaease/easegress/pull/590

**Pull request merged**